haystack/nodes/prompt/invocation_layer/sagemaker.py

```diff
@@ -0,0 +1,221 @@
+
+    # Pass sagemaker endpoint name and authentication details
+    pn = PromptNode(model_name_or_path="falcon-40b-my-sagemaker-inference-endpoint, model_kwargs={"profile_name": "my_aws_profile_name", "region_name": "eu-central-1"})
+    res = pn("what is the meaning of life?")
+    print(res)
+    """
+
+    def __init__(
+        self,
+        model_name_or_path: str,
+        max_length: int = 100,
+        aws_access_key_id: Optional[str] = None,
+        aws_secret_access_key: Optional[str] = None,
+        aws_session_token: Optional[str] = None,
+        region_name: Optional[str] = None,
+        profile_name: Optional[str] = None,
+        **kwargs,
+    ):
+        """
+        Instantiates the session with SageMaker using IAM based authentication via boto3.
+
+        :param model_name_or_path: The name for SageMaker Model Endpoint.
+        :param max_length: The maximum length of the output text.
+        :param aws_access_key_id: AWS access key ID.
+        :param aws_secret_access_key: AWS secret access key.
+        :param aws_session_token: AWS session token.
+        :param region_name: AWS region name.
+        :param profile_name: AWS profile name.
+        """
+        boto3_import.check()
+        super().__init__(model_name_or_path)
+        try:
+            session = boto3.Session(
+                aws_access_key_id=aws_access_key_id,
+                aws_secret_access_key=aws_secret_access_key,
+                aws_session_token=aws_session_token,
+                region_name=region_name,
+                profile_name=profile_name,
+            )
+            self.client = session.client("sagemaker-runtime")
+        except Exception as e:
+            raise SageMakerInferenceError(
+                f"Could not connect to SageMaker Inference Endpoint {model_name_or_path}."
+                f"Make sure the Endpoint exists and AWS environment is configured. {e}"
+            )
+        self.max_length = max_length
+
+        # for a list of supported parameters (TODO: verify if all of them are supported in sagemaker)
+        self.model_input_kwargs = {
+            key: kwargs[key]
+            for key in [
+                "best_of",
+                "details",
+                "do_sample",
+                "max_new_tokens",
+                "max_time",
+                "model_max_length",
+                "num_return_sequences",
+                "repetition_penalty",
+                "return_full_text",
+                "seed",
+                "temperature",
+                "top_k",
+                "top_p",
+                "truncate",
+                "typical_p",
+                "watermark",
+            ]
+            if key in kwargs
+        }
+
+        # We pop the model_max_length from the model_input_kwargs as it is not sent to the model
+        # but used to truncate the prompt if needed
+        model_max_length = self.model_input_kwargs.pop("model_max_length", 1024)
```

Comment:
Have you thought about inferring the `model_max_length` from `model_name_or_path`? For example, Haystack does that for ChatGPT models. I understand that we need to deal with any model from the model hub here. Instead of the default 1024, maybe a feature in future could be to automatically retrieve that setting for the model from the model hub? What do you think?

Reply:
That would be a great thing to do, except these fields are not always maintained in the model config files. Starting with flan models, sequence length size is dependent on available memory. Things are then complicated on top of this because now some LLMs reuse tokenizers from other models (tokenizers held this value in `tokenizer.model_max_length`). The most reasonable solution seems to be to start with a `model_max_length` of 1024, then document this parameter and let the users know that they should check the sequence length themselves and set it appropriately.